src/btree/bt_walk.c

```diff
@@ -624,3 +625,241 @@ __wti_tree_walk_skip(WT_SESSION_IMPL *session, WT_REF **refp, uint64_t *skipleaf
 
     return (0);
 }
+
+/*
+ * !!!
+ * __wt_page_npos --
+ *     Get the page's normalized position in the tree.
+ *     - If `path_str_offsetp` is set, return a string representation of the page's path.
+ *     - `start` is a position within the leaf page: 0 .. 1.
+ *       * When calculating a leaf page's position, use 0.5 to get the middle of the page.
+ *       * 0 and 1 are corner cases and can lead you to an adjacent page.
+ *       * Numbers outside of 0 .. 1 range will lead you to a prev/next page.
+ */
+double
+__wt_page_npos(WT_SESSION_IMPL *session, WT_REF *ref, double start, char *path_str,
+  size_t *path_str_offsetp, size_t path_str_sz_max)
+{
+    WT_PAGE_INDEX *pindex;
+    double npos;
+    uint32_t entries, slot;
+    int unused = 1; /* WT_UNUSED(snprintf) is screwed in GCC */
+
+    npos = start;
+    if (path_str)
+        *path_str_offsetp = 0;
+
+    WT_ENTER_PAGE_INDEX(session);
+    while (!__wt_ref_is_root(ref)) {
+        slot = UINT32_MAX; /* We get this invalid value in case of error */
+        __ref_index_slot(session, ref, &pindex, &slot);
+        entries = pindex->entries;
+        /*
+         * Depending on the implementation, `__ref_index_slot` might return an error or `slot`
+         * outside of range. Check for `slot < entries` ensures that it's a valid number. If it's
+         * not, then just skip the adjustment: the resulting number will be wrong but still within
+         * the range of the current page. Alternatively, could assign it any "reasonable" estimate
+         * like 0.5 or 0 / 1 depending on the walk direction.
+         */
+        if (slot < entries)
+            npos = (slot + npos) / entries;
+        if (path_str)
+            WT_UNUSED(unused = __wt_snprintf_len_incr(&path_str[*path_str_offsetp],
+                        path_str_sz_max - *path_str_offsetp, path_str_offsetp,
+                        "[%" PRIu32 "/%" PRIu32 "]", slot, entries));
+        __ref_ascend(session, &ref, NULL, NULL);
+    }
+    WT_LEAVE_PAGE_INDEX(session);
+
+    if (path_str) {
+        path_str[*path_str_offsetp] = 0;
+    }
+
+    return (WT_CLAMP(npos, 0.0, 1.0));
+}
+
+/*
+ * !!!
+ * __find_closest_page --
+ *     Find the closest suitable page according to flags.
+ *     - It should not be deleted.
+ *     - If WT_READ_CACHE is set, the page should be in memory.
+ */
+static int
+__find_closest_page(WT_SESSION_IMPL *session, WT_REF **refp, uint32_t flags)
+{
+    if (*refp == NULL || F_ISSET(*refp, WT_REF_FLAG_LEAF))
```

Comment:
Can we have a comment explaining `*refp == NULL` indicates we're on the root page?
Also how can we distinguish between NULL meaning a root page and NULL meaning we don't have a valid page?